### PR TITLE
feat(dashboard): add function permissions settings

### LINKED ISSUE #3468: dashboard: non-table objects
There are non-table objects that might be interesting to show in the database view:
- functions that return objects
- functions that work on objects
- materialized views
- others...

### PR DESCRIPTION
**Pull request by dbm03**
Resolves #3468

### Checklist

- [X] No breaking changes
- [X] Tests pass
- [X] New features have new tests
- [X] Documentation is updated (if applicable)
- [X] Title of the PR is in the correct format (see below)

### **PR Type**
Enhancement

___

### **Description**
- Add function permissions management UI: role-based grant/revoke for database functions via Hasura metadata API (both platform and local migration paths).
- Enable "Edit Permissions" action in the sidebar context menu for functions, gated by tracked-state awareness via the existing `useIsTrackedFunction` hook.
- Display permission state as full/partial/no-access based on explicit function permissions and inferred SELECT permissions on the return table, with special handling for the `inferFunctionPermissions` Hasura setting.
- Extend the OpenAPI spec and regenerate Hasura API types for `pg_create_function_permission` and `pg_drop_function_permission` bulk operations and migration steps.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  subgraph Sidebar
    DOA["DatabaseObjectActions"] -->|"Edit Permissions"| DBLI["DatabaseObjectListItem"]
  end

  subgraph "Permissions Drawer"
    EFPF["EditFunctionPermissionsForm"]
    EFPF --> FPD["FunctionPermissionsDescription"]
    EFPF --> PL["PermissionsLegend"]
  end

  subgraph Hooks
    EFPF --> UFPQ["useFunctionPermissionQuery"]
    EFPF --> UMFPM["useManageFunctionPermissionMutation"]
    UFPQ --> UEM["useExportMetadata"]
    UMFPM -->|platform| MFP["manageFunctionPermission"]
    UMFPM -->|local| MFPM["manageFunctionPermissionMigration"]
  end

  subgraph Utils
    EFPF --> GFPS["getFunctionPermissionState"]
  end

  DBLI -->|"onEditPermissions"| UDA["useDataBrowserActions"]
  UDA -->|"openDrawer"| EFPF
  DOA --> UITF["useIsTrackedFunction"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody>

<tr><td><strong>Sidebar integration</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>DatabaseObjectActions.tsx</strong><dd><code>Enable permissions menu item for functions; add useIsTrackedFunction check</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-d9b3d469e71dcc60c958dc6dca4581f2dbc9246ab72cea6129f5c9e2b7665fb1">+21/-14</a></td>
</tr>
<tr>
  <td><strong>DatabaseObjectListItem.tsx</strong><dd><code>Route function permission clicks to handleEditFunctionPermission</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-2e4cb8aeaebe85f4bf5c23241df2e301220c306829c038b82285258dc24a3a80">+13/-6</a></td>
</tr>
</table></details></td></tr>

<tr><td><strong>Permissions UI components</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>EditFunctionPermissionsForm.tsx</strong><dd><code>New form: role-based permission grid with collapsible toggle per role</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-f859d1d9f373ae0d0faf90af82c0a70b391d8083ba06fee6a4e4909bb9e672af">+440/-0</a></td>
</tr>
<tr>
  <td><strong>FunctionPermissionsDescription.tsx</strong><dd><code>Informational text explaining permission inheritance from return table SELECT</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-9e4c180a1ff71def62580338cce8f3ae03e733243aba6fb669958d3076c88e40">+68/-0</a></td>
</tr>
<tr>
  <td><strong>PermissionsLegend.tsx</strong><dd><code>Reusable legend showing full/partial/no access icons</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-ccfb2f4ca383a1ee1691552732557a740ae49d5c0636acc06d26398b4fd364c3">+21/-0</a></td>
</tr>
<tr>
  <td><strong>index.ts (2 files)</strong><dd><code>Barrel exports for new components</code></dd></td>
  <td>+2/-0</td>
</tr>
</table></details></td></tr>

<tr><td><strong>Data browser actions hook</strong></td><td><details><summary>1 file</summary><table>
<tr>
  <td><strong>useDataBrowserActions.tsx</strong><dd><code>Add handleEditFunctionPermission to open the permissions drawer</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-9cb58e1886d2992111cded05c73a266c5808dc857ef4dd87a7b749cc1f27c50b">+44/-0</a></td>
</tr>
</table></details></td></tr>

<tr><td><strong>Permission query & mutation hooks</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>useFunctionPermissionQuery.ts</strong><dd><code>Derive function permissions from exported metadata via useExportMetadata</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-0f8ca6557feae6f9b36c6458ad1d6bfb017644d1611913015584a5286943660f">+47/-0</a></td>
</tr>
<tr>
  <td><strong>useManageFunctionPermissionMutation.ts</strong><dd><code>Mutation hook branching between platform metadata API and local migration</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-cfaec613337e7780eab5b18dbf7c8ea4c5d536c8f94a07b9eb5eefa2f2ab317c">+82/-0</a></td>
</tr>
<tr>
  <td><strong>manageFunctionPermission.ts</strong><dd><code>Platform path: bulk metadata operation for create/drop function permission</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-24b1dd44a5522d761e3c841372e98d79f39765024829ca92d85567e24c92ffd6">+54/-0</a></td>
</tr>
<tr>
  <td><strong>manageFunctionPermissionMigration.ts</strong><dd><code>Local path: migration with up/down steps for create/drop function permission</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-08d290aaeed3abe24b34912e461e9cfd42f1e56215d969873604f84b5534e589">+68/-0</a></td>
</tr>
<tr>
  <td><strong>index.ts (2 files)</strong><dd><code>Barrel exports for new hooks</code></dd></td>
  <td>+3/-0</td>
</tr>
</table></details></td></tr>

<tr><td><strong>Utility functions</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>getFunctionPermissionState.ts</strong><dd><code>Pure function deriving allowed/partial/not-allowed from infer setting + permissions</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-8780484beec65acf97255a99f282389050d4484fc25d3726cf7b1c23fd03c2e3">+27/-0</a></td>
</tr>
<tr>
  <td><strong>getFunctionPermissionState.test.ts</strong><dd><code>Unit tests covering inferred, explicit, mutation, and edge-case permission states</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-db540b102320350251ef40f1f9a5ecf54c14ec4431c82306ffec8b1b6222eaaf">+97/-0</a></td>
</tr>
<tr>
  <td><strong>index.ts</strong><dd><code>Barrel export for getFunctionPermissionState</code></dd></td>
  <td>+5/-0</td>
</tr>
</table></details></td></tr>

<tr><td><strong>Generated Hasura API schemas</strong></td><td><details><summary>14 files</summary><table>
<tr>
  <td><strong>openapi.yaml</strong><dd><code>OpenAPI spec additions for function permission operations and metadata</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-8f0d77ee7586b6cb667194c6a05a63ece4f68309522828d3402a96b54f75d16f">+123/-0</a></td>
</tr>
<tr>
  <td><strong>functionPermission.ts</strong><dd><code>Generated FunctionPermission type (role string)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-63adb9fd9c7ef70e22de72085eed8697f9b8226c323ba314487d0d50e0d829e6">+12/-0</a></td>
</tr>
<tr>
  <td><strong>create/drop FunctionPermission*.ts (8 files)</strong><dd><code>Generated types for bulk operations, steps, args, and step types</code></dd></td>
  <td>+128/-0</td>
</tr>
<tr>
  <td><strong>exportMetadataResponseMetadataSourcesItemFunctionsItem.ts</strong><dd><code>Add permissions array to function metadata item</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-1234bd01781b6c7fb4f95149b029e90c8a159d1fadfe18838d14e7ed8e649cc0">+3/-0</a></td>
</tr>
<tr>
  <td><strong>index.ts</strong><dd><code>Re-export new generated schemas</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-eefb97472cd23c251a23da804bcaf5d8940231e3dccabf72d4459985b34c94c0">+11/-0</a></td>
</tr>
<tr>
  <td><strong>metadataOperation.ts</strong><dd><code>Add function permission bulk operations to MetadataOperation union</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-662721b9808d797951e7556fabefe725c8a38eacf7fcbc95d71a4708411b56c4">+4/-0</a></td>
</tr>
<tr>
  <td><strong>migrationStep.ts</strong><dd><code>Add function permission steps to MigrationStep union</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3995/files#diff-2a3b0d700c799be555b8c3dca513107d6eec6ba11dea63f3f276a01b1e162a0e">+5/-1</a></td>
</tr>
</table></details></td></tr>

</tbody></table>

</details>

___

### CODE REVIEW

File: dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserSidebar/DatabaseObjectActions.tsx [hunk 1]
```diff
@@ -118,18 +127,16 @@ function DatabaseObjectActions({
...
-            <Users className="h-4 w-4" /> <span>Edit Permissions</span>
-          </DropdownMenuItem>
-        )}
+        <DropdownMenuItem
+          className={cn(menuItemClassName, {
+            'italic opacity-50 hover:cursor-default hover:bg-transparent':
+              !isTracked,
+          })}
+          disabled={!isTracked}
+          onClick={isTracked ? onEditPermissions : undefined}
```
Comment (robertkasza) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserSidebar/DatabaseObjectActions.tsx]: This smells a little bit for me. In `DatabaseObjectListItem` we always pass a function and then in the child we decide whether we are using that function or not. And we even doing the distinction between functions and other objects, but not checking if they are tracked or not. We even have the information about tracking in the parent component.

Comment (dbm03) [reply dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserSidebar/DatabaseObjectActions.tsx]: I've updated this part so I pass `isUntracked` as props from the parent to the child component. Now the child doesn't have to check whether it's tracked by itself.

Please check, I think it's cleaner now
